# Tugas UTS

Nama : Achmad Ilham Firmansyah

NIM : 210411100127

Lakukan analisa data klasifikasi dengan sumber data https://github.com/kb22/Heart-Disease-Prediction/blob/master/dataset.csv dengan menggunakan naive bayes dan knn untuk data.

## 1. Metode Naive Bayes

## Mengimport package pandas dan numpy. Menginputkan data yang akan dianalisis.

In [2]:
import pandas as pd
import numpy as np
data = pd.read_csv('https://raw.githubusercontent.com/kb22/Heart-Disease-Prediction/master/dataset.csv')
data

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

## Menampilkan informasi data

In [3]:
# menampilkan informasi data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [4]:
data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

## Melakukan pengecekan deret kosong.


In [5]:
data.empty

False

Output menunjukkan false berarti tidak ada deret kosong.

## Mengecek ukuran data


In [6]:
data.size

4242

Ukuran data yaitu sebanyak 4242.

## Menentukan variabel independen dan dependen.
Kolom target dihapus dari data karena akan menjadi variabel dependen.

In [7]:
# Variabel independen
x = data.drop(['target'], axis = 1)
x.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [8]:
Y = data['target'].values
Y[0:61]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

In [9]:
# Variabel dependen
y = data['target']
y.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

## Membagi dataset menjadi training set dan test set.
Dataset dibagi menjadi 80% training set dan 20% test set.

In [10]:
# Import train_test_split function
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 123)

## Menentukan prediksi pada training set dan test set.

In [11]:
# Import Gaussian Naive Bayes model
from sklearn.naive_bayes import GaussianNB
# Mengaktifkan/memanggil/membuat fungsi klasifikasi Naive bayes
modelnb = GaussianNB()
# Memasukkan data training pada fungsi klasifikasi naive bayes
nbtrain = modelnb.fit(x_train, y_train)
nbtrain.class_count_

array([108., 134.])

#Hasil prediksi dari x_test

In [12]:
# Menentukan hasil prediksi dari x_test
y_pred = nbtrain.predict(x_test)
y_pred

array([1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0])

## Menentukan nilai probabilitas x_test

In [13]:
# Menentukan probabilitas hasil prediksi
nbtrain.predict_proba(x_test)

array([[5.82571613e-04, 9.99417428e-01],
       [9.99240610e-01, 7.59390366e-04],
       [9.99938869e-01, 6.11309321e-05],
       [9.99992658e-01, 7.34177465e-06],
       [3.63366729e-03, 9.96366333e-01],
       [9.99951069e-01, 4.89308903e-05],
       [8.58720034e-03, 9.91412800e-01],
       [5.98819913e-01, 4.01180087e-01],
       [5.96228807e-04, 9.99403771e-01],
       [9.33511439e-01, 6.64885614e-02],
       [9.97942318e-01, 2.05768198e-03],
       [4.57117589e-01, 5.42882411e-01],
       [9.99999986e-01, 1.36224796e-08],
       [9.99995730e-01, 4.26988712e-06],
       [5.16991611e-01, 4.83008389e-01],
       [1.00000000e+00, 5.08353833e-11],
       [9.10174530e-01, 8.98254697e-02],
       [9.86808541e-01, 1.31914589e-02],
       [5.72076923e-01, 4.27923077e-01],
       [2.24348439e-01, 7.75651561e-01],
       [1.35283843e-01, 8.64716157e-01],
       [9.99163302e-01, 8.36698360e-04],
       [4.11827050e-05, 9.99958817e-01],
       [9.82142245e-01, 1.78577552e-02],
       [9.874878

#Confussion Matrix
Prediksi target dinyatakan gagal dan ternyata target gagal sebanyak 27, prediksi target gagal dan ternyata target berhasil 7, prediksi traetment target dan ternyata target gagal sebanyak 3 dan prediksi target berhasil dan ternyata target berhasil sebanyak 24.

In [14]:
# import confusion_matrix model
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[27,  3],
       [ 7, 24]])

#Perhitungan nilai akurasi

In [15]:
# Menghitung nilai akurasi dari klasifikasi naive bayes 
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.79      0.90      0.84        30
           1       0.89      0.77      0.83        31

    accuracy                           0.84        61
   macro avg       0.84      0.84      0.84        61
weighted avg       0.84      0.84      0.84        61



.

# 2. Metode KNN

In [16]:
import pandas as pd
data = pd.read_csv('https://raw.githubusercontent.com/kb22/Heart-Disease-Prediction/master/dataset.csv')
data

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

##Menampilkan data

In [17]:
data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

##Mengecek ukuran data

In [18]:
data.shape

(303, 14)

##Menjadikan dataset sebagai input (X) dan target (Y)

In [19]:
#Membuat dataframe dengan semua training data kecuali kolom target
X = data.drop(columns=["target"])
#mengecek apakah variabel target sudah dihapus
X.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [20]:
Y = data["target"].values
Y[0:150]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

##Membagi dataset menjadi training set dan test set.

In [21]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=1, stratify=Y)

##Scaling data

In [22]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
scaler.fit(X_train)

X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

##Membuat testing model

In [23]:
from sklearn.neighbors import KNeighborsClassifier
# Membuat KNN classifier
knn = KNeighborsClassifier(n_neighbors = 3)
knn.fit(X_train_scaled,Y_train)

KNeighborsClassifier(n_neighbors=3)

##Menampilkan 5 model prediksi pertama pada data test

In [24]:
knn.predict(X_test_scaled)[0:5]

array([1, 1, 1, 1, 1])

##Akurasi model dari data test

In [25]:
knn.score(X_test_scaled, Y_test)

0.8032786885245902